main.rb

```diff
@@ -0,0 +1,117 @@
+# rubocop:disable Metrics/CyclomaticComplexity
+# rubocop:disable Metrics/PerceivedComplexity
+
+module Enumerable
+  def my_each
+    counter = 0
+    while counter < size
+      if instance_of(Hash) || instance_of?(Range)
+        new_hash = to_a
+        yield(new_hash[counter])
+      else
+        yield(self[counter])
+      end
+      counter += 1
+    end
+  end
+
+  def my_each_with_index
+    counter = 0
+    while counter < size
+      if instance_of(Hash) || instance_of?(Range)
+        new_hash = to_a
+        yield(new_hash[counter], counter)
+      else
+        yield(self[counter], counter)
+      end
+      counter += 1
+    end
+  end
+
+  def my_select
+    if instance_of?(Hash)
+      new_hash = {}
+      my_each { |key, value| new_hash[key] = value if yield(key, value) }
+      new_hash
+    else
+      arr = []
+      my_each { |item| arr.push(item) if yield(item) }
+      arr
+    end
+  end
+
+  def my_all?(type = nil)
+    if block_given?
+      my_each { |item| return false if yield(item) == false }
+    elsif type.nil?
+      my_each { |item| return false if item.nil? }
+    else
+      my_each { |item| return false if item.class != type && item.class.superclass != type }
+    end
+    true
+  end
+
+  def my_any?(type = nil)
+    if block_given?
+      my_each { |item| return true if yield(item) == true }
+    elsif type.nil?
+      my_each { |item| return true if item.nil? }
+    else
+      my_each { |i| return true if i.instance_of?(type.class) || i.class.superclass == type || i.is_a?(type) }
+    end
+    false
+  end
+
+  def my_none?(type = nil)
+    if block_given?
+      my_each { |item| return false if yield(item) == true }
+    elsif type.nil?
+      my_each { |item| return false if item.nil? && item != false }
+    else
+      my_each { |item| return false if item.instance_of?(type) || item.class.superclass == type || item.is_a?(type) }
+    end
+    true
+  end
+
+  def my_count(count = nil)
+    counter = 0
+    if block_given?
+      my_each { |item| counter += 1 if yield(item) == true }
+    elsif count.nil?
+      my_each { counter += 1 }
+    else
+      my_each { |item| counter += 1 if item == count }
+    end
+    counter
+  end
+
+  def my_map
+    temp_arr = []
+    my_each { |item| temp_arr.push(yield(item)) unless yield(item) == false }
+    temp_arr
+  end
```

Comment:
#### my_map
- [ ] 20. #my_map returns a new array with the results of running a given block once for every element in enum.
- [ ] 21. #my_map returns a new array with the results of running a given block once for every element in enum::range
- [ ] 22. #my_map returns an Enumerator if no block is given
- [ ] 23. #my_map accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.
- [ ] 24. #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.